5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_comparisons.py

```diff
@@ -0,0 +1,278 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# ## Perform bootstrapping ROC AUC method to determine if QC significantly improves performance of model
+# 
+# In this method, we have trained two models; one with QC'd data and the other without QC'd data (more noise). 
+# To assess quality, apply the models to a holdout dataset that both models have never seen that has been QC'd.
+# We take subsamples of this dataset to perform a bootstapping method that will calculate the ROC AUC across the QC'd holdout samples.
+# 
+# The goal of this analysis is to see how well the models perform when applied to a new dataset that has only "good" cells, since we will be applying QC to all future plates.
+# 
+
+# ## Import libraries
+
+# In[1]:
+
+
+import pathlib
+import joblib
+from joblib import load
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+from sklearn.metrics import roc_auc_score
+from sklearn.utils import resample
+import matplotlib.colors as mcolors
+from scipy.stats import ttest_ind
+
+import sys
+
+sys.path.append("../../utils")
+from training_utils import get_X_y_data
+
+
+# ## Define function
+
+# In[2]:
+
+
+# Define function to preform bootstrapping
+def bootstrap_roc_auc(y_true, y_pred, n_bootstraps=1000):
+    """
+    Perform bootstrapping to compute the distribution of ROC AUC scores.
+
+    This function generates a bootstrapped distribution of ROC AUC scores by 
+    resampling the provided true labels and predicted probabilities with 
+    replacement.
+
+    Parameters:
+    ----------
+    y_true : array-like of shape (n_samples,)
+        True binary labels (0 or 1) for the dataset.
+        
+    y_pred : array-like of shape (n_samples,)
+        Predicted probabilities or scores for the positive class.
+        
+    n_bootstraps : int, optional, default=1000
+        Number of bootstrap iterations to perform.
+
+    Returns:
+    -------
+    bootstrapped_scores : np.ndarray
+        An array of bootstrapped ROC AUC scores. Each element represents the 
+        ROC AUC computed for a resampled dataset.
+    """
+
+    bootstrapped_scores = []
+    for i in range(n_bootstraps):
+        indices = resample(np.arange(len(y_true)), replace=True)
+        if len(np.unique(y_true[indices])) < 2:
+            continue
```

Comment:
I'm not sure if this if statement actually is doing anything. There's no `else` part to direct the logic